RNMobile - Image Block: Add caption. (v2)

etoledom opened the pull request:
## Description
This PR adds captions to the mobile version of the Image Block.

![caption](https://user-images.githubusercontent.com/9772967/45844866-fa7a5580-bcf9-11e8-8450-5a9f93adb4a8.gif)

**To test:**
- Please checkout the `gutenberg-mobile` project and follow the instructions.

etoledom committed:
RNMobile Image Block: Adding image caption.

etoledom committed:
RNMobile Image Block: Fixing lint issues.

etoledom commented:
Those commits from @mzorz are already on `master`. Not sure why are appearing again here. 🤔

etoledom commented:
Thank you @hypest ! I believe now it's ready for review.

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -39,6 +40,44 @@ export default function ImageEdit( { attributes } ) {
+		return null;
+	}
+
+	const textContainer = ( children ) => {
```
Any particular reason we fall back to plain textview instead of just letting the editing component in?

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -39,6 +40,44 @@ export default function ImageEdit( { attributes } ) {
+	if ( isSelected ) {
+		return textContainer( textInput );
+	}
+	return textContainer( text );
```
Any particular reason we fall back to plain textview instead of just letting the editing component in?

etoledom replied:
I tried using the `editable={ isSelected }` property on `TextInput`. It had these issues:

When the block is **not** selected, with a caption written, and I tap over it to select the block, 
- **IOS**: The tap is consumed by the TextInput component and it doesn't select the block.
- **Android**: That property seems to not work and it starts editing the caption anyway, without actually selecting the block.

I tested this approach and it worked as expected, so I decided to keep it. If there's a better way I'll be happy to test it!

hypest replied:
Yeah, I think the focus/selection issues are already there for the rest of the blocks anyway and we will probably tackle them in unison.

Besides, the textview only solution "solves" the tap-to-focus-block but needs an extra tap to place the cursor.

I think leaving the TextInput (or RichText as the web side does) should be fine for this iteration and should simplify the PR diff for now. WDYT?

etoledom replied:
Sounds good!
And yes! it will get quite a lot simpler. 😁

etoledom committed:
RNMobile Image Block: Removing usage of `Text` component for Image Caption.
Now it uses just TextInput in both selected and not-selected state to simplify implementation.

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -39,6 +40,27 @@ export default function ImageEdit( { attributes } ) {
 	);
 }
+
+const CaptionTextInput = ( setAttributes, caption, isSelected ) => {
```
At this point it seems that `CaptionTextInput` just checks for caption presence/length besides rendering the caption elements. WDYT about copying what the web-side does and just check and inline the JSX snippet directly, [like this](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/image/edit.js#L488-L498)? This will bring us closer to the web-side and will be easier to compare what's different on the mobile.

etoledom replied:
I like that is more declarative in this way, but I agree on trying to be closer to the web-side! I'll push a commit with the change 👍

etoledom committed:
RNMobile Image Block: Removed `const CaptionTextInput` in favor of inline declaration.

hypest reviewed (approved):
LGTM!
